packages/firestore/test/util/values.ts

```diff
@@ -0,0 +1,116 @@
+import { DatabaseId } from '../../src/core/database_info';
+import { DocumentKey } from '../../src/model/document_key';
+import { fail } from '../../src/util/assert';
+import { Dict, forEach } from '../../src/util/obj';
+
+/** Test helper to create Firestore Value protos from JavaScript types. */
+
+// TODO(mrschmidt): Move into UserDataConverter
+export function valueOf(
+  input: unknown,
+  useProto3Json: boolean = false
+): api.Value {
+  if (input === null) {
+    return { nullValue: 'NULL_VALUE' };
+  } else if (typeof input === 'number') {
+    if (typeUtils.isSafeInteger(input)) {
+      return { integerValue: String(input) };
+    } else {
+      if (useProto3Json) {
+        // Proto 3 let's us encode NaN and Infinity as string values as
+        // expected by the backend. This is currently not checked by our unit
+        // tests because they rely on protobuf.js.
+        if (isNaN(input)) {
+          return { doubleValue: 'NaN' } as {};
+        } else if (input === Infinity) {
+          return { doubleValue: 'Infinity' } as {};
+        } else if (input === -Infinity) {
+          return { doubleValue: '-Infinity' } as {};
+        }
+      }
+      return { doubleValue: input };
+    }
+  } else if (typeof input === 'boolean') {
+    return { booleanValue: input };
+  } else if (typeof input === 'string') {
+    return { stringValue: input };
+  } else if (input instanceof Timestamp) {
+    return {
+      timestampValue: {
+        seconds: String(input.seconds),
+        nanos: input.nanoseconds
+      }
+    };
+  } else if (input instanceof GeoPoint) {
+    return {
+      geoPointValue: {
+        latitude: input.latitude,
+        longitude: input.longitude
+      }
+    };
+  } else if (input instanceof Blob) {
+    if (useProto3Json) {
+      return { bytesValue: input._byteString.toBase64() };
+    } else {
+      return { bytesValue: input._byteString.toUint8Array() };
+    }
+  } else if (input instanceof DocumentKeyReference) {
+    return {
+      referenceValue:
+        'projects/project/databases/(default)/documents/' + input.key.path
+    };
+  } else if (Array.isArray(input)) {
+    return {
+      arrayValue: { values: input.map(el => valueOf(el, useProto3Json)) }
+    };
+  } else if (typeof input === 'object') {
+    const result: api.Value = { mapValue: { fields: {} } };
+    forEach(input as Dict<unknown>, (key: string, val: unknown) => {
+      result.mapValue!.fields![key] = valueOf(val, useProto3Json);
+    });
+    return result;
+  } else {
+    fail(`Failed to serialize field: ${input}`);
+  }
+}
+
+/** Creates a MapValue from a list of key/value arguments. */
+export function mapOf(...entries: unknown[]): api.Value {
+  const result: api.Value = { mapValue: { fields: {} } };
+  for (let i = 0; i < entries.length; i += 2) {
+    result.mapValue!.fields![entries[i] as string] = valueOf(
+      entries[i + 1],
+      /* useProto3Json= */ false
```

Comment:
Curious: why is it OK to set it to always false here?

Reply:
For now, this is a test utility only. The best I can tell you that we will find out soon if it indeed is ok.